restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java

```diff
@@ -135,28 +144,367 @@ public ReactiveDocumentService(
+        JsonNode whereNode = objectMapper.readTree(where);
+        expression =
+            expressionParser.constructFilterExpression(
+                prependPath, whereNode, db.treatBooleansAsNumeric());
+      } catch (JsonProcessingException ex) {
+        throw new ErrorCodeRuntimeException(ErrorCode.DOCS_API_SEARCH_WHERE_JSON_INVALID);
+      }
+    }
+    return expression;
+  }
+
+  private Collection<List<String>> getFields(String fields) {
+    Collection<List<String>> fieldPaths = Collections.emptyList();
+    if (null != fields) {
+      try {
+        JsonNode fieldsNode = objectMapper.readTree(fields);
+        fieldPaths = DocsApiUtils.convertFieldsToPaths(fieldsNode);
+      } catch (JsonProcessingException ex) {
+        throw new ErrorCodeRuntimeException(ErrorCode.DOCS_API_SEARCH_FIELDS_JSON_INVALID);
+      }
+    }
+    return fieldPaths;
+  }
+
+  // we need to transform the stuff to support array elements
+  private List<String> processSubDocumentPath(List<String> subDocumentPath) {
+    // TODO after https://github.com/stargate/stargate/pull/1105 encode chars as well
+    return subDocumentPath.stream()
+        .map(path -> DocsApiUtils.convertArrayPath(path))
+        .collect(Collectors.toList());
+  }
+
   public ObjectNode createJsonMap(
-      DocumentDB db, List<RawDocument> docs, Collection<List<String>> fieldPaths) {
+      DocumentDB db,
+      List<RawDocument> docs,
+      Collection<List<String>> fieldPaths,
+      boolean writeAllPathsAsObjects) {
     ObjectNode docsResult = objectMapper.createObjectNode();
 
     for (RawDocument doc : docs) {
-      // filter needed rows only
-      List<Row> rows = doc.rows();
-      if (!fieldPaths.isEmpty()) {
-        rows =
-            doc.rows().stream()
-                .filter(
-                    row ->
-                        fieldPaths.stream()
-                            .anyMatch(fieldPath -> DocsApiUtils.isRowOnPath(row, fieldPath)))
-                .collect(Collectors.toList());
-      }
-
       // create document node and set to result
-      JsonNode node = jsonConverter.convertToJsonDoc(rows, false, db.treatBooleansAsNumeric());
+      JsonNode node =
+          documentToNode(doc, fieldPaths, writeAllPathsAsObjects, db.treatBooleansAsNumeric());
       docsResult.set(doc.id(), node);
     }
 
     return docsResult;
   }
+
+  public ArrayNode createJsonArray(
+      DocumentDB db,
+      List<RawDocument> docs,
+      Collection<List<String>> fieldPaths,
+      boolean writeAllPathsAsObjects) {
+    ArrayNode docsResult = objectMapper.createArrayNode();
+
+    for (RawDocument doc : docs) {
+      // create document node and set to result
+      JsonNode node =
+          documentToNode(doc, fieldPaths, writeAllPathsAsObjects, db.treatBooleansAsNumeric());
+
+      // TODO Confirm: adding empty nodes to the results array makes no sense?
+      //  should be revisited
+      if (!node.isEmpty()) {
+        docsResult.add(node);
+      }
```

Comment:
+1 to exclude empty nodes. @EricBorczuk : WDYT?

Reply:
Eric is on the holiday, I would also vote for the removing the empty nodes, OK?

Reply:
ok, let's remove